### PR TITLE
feat: refresh keys when current jwt kid is not found

### PR DESCRIPTION
**Pull request by alexzhangreslv**
I find that currently keys will only be fetched via JWKS endpoints every 15 minutes. Thus if a JWT token is created with a new key, it will be validated failed until the next background refresh.
So I have made the following changes:

1. If JWKS endpoint list is not empty, when a key ID is not found in the keys map, the background keys refresh will be forced to trigger before proceeding token verification.
2. A read-write lock is added to the `keys` map since it could be accessed by two goroutines.

### CODE REVIEW

**Comment by eshepelyuk** [comment]
Hello, sorry for being late with review :)

Got one doubtingpoint - please confirm is my understanding is correct.

So, every time that kay is not found in cache - the call to remote is made to refresh the keys ?

**Comment by alexzhangreslv** [comment]
> Hello, sorry for being late with review :)
> 
> Got one doubt point, so confirm if my understanding is correct.
> 
> This PR does following: every time jwt key is not found in cache - the call to remote is made to refresh the keys ?

Yes. And the next automatic background refresh will be reset as 15 min later.

**Comment by eshepelyuk** [comment]
> > Hello, sorry for being late with review :)
> > Got one doubt point, so confirm if my understanding is correct.
> > This PR does following: every time jwt key is not found in cache - the call to remote is made to refresh the keys ?
> 
> Yes. And the next automatic background refresh will be reset as 15 min later.

But doesn't it mean that if there will be a lot of "erroneous" requests, or malicious traffic with errors - than the downstream JWKS server will be flooded with such requests ?

**Comment by alexzhangreslv** [comment]
> > > Hello, sorry for being late with review :)
> > > Got one doubt point, so confirm if my understanding is correct.
> > > This PR does following: every time jwt key is not found in cache - the call to remote is made to refresh the keys ?
> > 
> > 
> > Yes. And the next automatic background refresh will be reset as 15 min later.
> 
> But doesn't it mean that if there will be a lot of "erroneous" requests, or malicious traffic with errors - than the downstream JWKS server will be flooded with such requests ?

Yes it will, yet I assume it is the potential risk a JWKS server is responsible to take. However, if you really concern about it, I can later update the branch by applying rate limit here, e.g. to guarantee that the JWKS server would never be called more than twice per second. Will that be acceptable?

**Comment by eshepelyuk** [comment]
I suggest instead to make this functionality conditional, i.e. driven by q new flag. So it could he responsibility of plugin users.
The idea is to avoid unnecessary code complication.

**Comment by alexzhangreslv** [comment]
Sure, I have added a new flag ForceRefreshKeys to turn on/off this function.

File: jwt.go [hunk 1]
```diff
@@ -189,20 +194,43 @@ func New(_ context.Context, next http.Handler, config *Config, _ string) (http.H
...
+			if config.ForceRefreshKeys {
+				jwtPlugin.forceRefreshCmd = make(chan chan<- struct{})
+			}
 			go jwtPlugin.BackgroundRefresh()
 		}
 	}
 	return jwtPlugin, nil
 }
 
 func (jwtPlugin *JwtPlugin) BackgroundRefresh() {
```
Comment (eshepelyuk) [inline jwt.go:206]: Can you plz explain how exactly this function works now. I don't understand golang code that well.

**Review by eshepelyuk — COMMENTED** [review]
Plz rebase and squash this PR into single commit.

File: jwt.go (hunk 1 above)
Comment (alexzhangreslv) [reply jwt.go:206]: Sure. When this `ForceRefreshKeys` is set true, the channel `forceRefreshCmd` will be created for _**main routine**_ to send "Force Refresh" command to _**background refresh routine**_. The message of this channel (`make(chan [CHANNEL_MESSAGE_TYPE])`) is a one-time sending-only channel (`chan<-`) of an empty struct (`struct{}`) used as callback from the _**background refresh routine**_.

So for each of the two routines, the procedures are like:

_**Background Refresh Routine**_:
1. Listen to the channel `forceRefreshCmd` (in addition to a 15-min-timeout channel)
2. Receive a message (which is a one-time callback channel) from the channel `forceRefreshCmd`, indicating keys should be refreshed immediately
3. Refresh keys
4. Send an empty struct to the callback channel, in order to notify _**main routine**_ that keys are already refreshed

_**Main Routine**_ (when key is not found):
1. Create a new callback channel
2. Send the callback channel as a message to channel `forceRefreshCmd`, in order to notify that _**background refresh routine**_ should be refreshed immediately
3. Listen to the callback channel created on step 1
4. Receive a message of an empty struct, indicating that keys have been refreshed just now
5. Retry finding key ID of current JWT token and proceed

**Comment by alexzhangreslv** [comment]
> Plz rebase and squash this PR into single commit.

Done

**Pull request merged**